text/0000-dyn-type.md

```diff
@@ -0,0 +1,2261 @@
+        "dyn" "type"
+        name:IDENT
+        generics:Generics?
+        "(" custom_dst_kind:CustomDstKind ")"
+        where_clause:WhereClause?
+        "=" ty:Type ";"
+    };
+
+CustomDstKind =
+    "regular" ":" ty:Type |
+    "inline";
+```
+
+Syntactical examples:
+
+```rust ,ignore
+#[repr(C)]
+pub dyn type CStr(inline) = [u8];
+
+dyn type Mat<T>(regular: MatMeta) = [T];
+
+dyn type Thin<P: ?Sized>(inline)
+where
+    P: Aligned,
+= (P::Metadata, P);
+```
+
+### Semantics
+
+When creating a *regular DST*, the compiler will automatically generate:
+
+```rust ,ignore
+dyn type RegDst<T, U>(regular: RegMeta<T, U>)
+where
+    T: Bounds<U>,
+= ReducedTy<T, U>;
+
+// implies:
+
+impl<T, U> core::marker::Pointee for RegDst<T, U>
+where
+    T: Bounds<U>,
+{
+    type Metadata = RegMeta<T, U>;
+}
+
+unsafe impl<T, U> core::dst::Derived for RegDst<T, U>
+where
+    T: Bounds<U>,
+{
+    type Reduced = ReducedTy<T, U>;
+    unsafe fn reduce_metadata(val: *const Self) -> Self::Reduced::Metadata {
+        let meta = core::ptr::metadata(val);
+        <Self as core::dst::Regular>::reduce_with_metadata(meta)
+    }
+}
+```
+
+When creating an *inline DST*, the compiler will automatically generate:
+
+```rust ,ignore
+dyn type InlineDst<T, U>(inline)
+where
+    T: Bounds<U>,
+= ReducedTy<T, U>;
+
+// implies:
+
+impl<T, U> core::marker::Pointee for InlineDst<T, U>
+where
+    T: Bounds<U>,
+{
+    type Metadata = ();
+}
+
+unsafe impl<T, U> core::dst::Derived for InlineDst<T, U>
+where
+    T: Bounds<U>,
+{
+    type Reduced = ReducedTy<T, U>;
+    unsafe fn reduce_metadata(val: *const Self) -> Self::Reduced::Metadata {
+        <Self as core::dst::Inline>::reduce_with_ptr(val as *const ())
+    }
+}
+```
+
+Users are expected to `core::dst::Regular` and `core::dst::Inline` themselves.
+Failing to do so should result in the normal [E0277] *“the trait bound is not satisfied”* error at the declaration site.
+
+```
+error[E0277]: the trait bound `CStr: std::dst::Inline` is not satisfied
+  --> src/lib.rs:11:9
+   |
+11 | pub dyn type CStr(inline) = [u8];
+   |                   ^^^^^^ the trait `std::dst::Inline` is not implemented for `CStr`
+```
+
+[E0277]: https://doc.rust-lang.org/error-index.html#E0277
+
+The variance of the generic parameters `T`, `U` will be the same as the tuple `(Self::Reduced, Self::Metadata)`.
+
+If `Self::Reduced` implements an auto-trait (e.g. `Send`/`Sync`), the custom DST itself will also implement it
+(`Self::Metadata` is not considered).
+
+The `#[repr]` attributes can be applied on custom DSTs.
+
+* With `#[repr(align(n))]`, the `align_of` and `align_of_val` methods will always return `n`,
+    and `Aligned` is always implemented regardless of `Self::Reduced`.
+* With `#[repr(C)]` on an inline DST, the type is FFI-safe.
+    `#[repr(C)]` on regular DST will cause error [E0517] *“attribute should be applied to struct, enum or union”*.
+* Everything else (packed, simd, transparent) causes [E0517] error.
```

Comment:
I'm surprised by `transparent`; could you go into why?

Reply:
The `#[repr(transparent)]` is no-op. Added an option that allows it to be ignored.